### DISCUSSION TITLE
Visualizing extracted network and tiff image together in Paraview

### DISCUSSION THREAD

**Discussion by nuritg** [PMEAL/porespy#708] [Q&A]
Hello,
I got the 3D network but when I'm trying to save it by:

```python
ps.io.to_vtk(np.array(m3d, dtype=int)[:, :, :], '3d-test')
```
I'm getting the error:
```python
OSError: [Errno 22] Invalid argument: '3d-test.vti'
```
(this line: `prj.export_data(filename= '3d-try', filetype='vtk')`, seems to work fine). 

Thanks!

**Comment by nuritg** [comment]
I think my question was not clear, so I will try to make it better.
after running the code below:
```
snow_output = ps.networks.snow2(
    m3d[:,:,:],
    voxel_size=1,
)
net = ps.networks.regions_to_network(snow_output.regions, voxel_size=1)
pn = op.network.GenericNetwork()
pn.update(net)

```
I get only vtp file in the Paraview, and working on it gives me that result:
![image](https://user-images.githubusercontent.com/36815224/183401340-8ba79ed9-3c18-4915-9ab2-8f4b33f297fa.png)

Is there something that I should do differently? 
Thanks a lot!

**Reply by jgostick** [reply]
Judging by your colorbar you have pore sizes spaning 6 orders of magnitude, so maybe they are present but not visible.  You can try to remove scaling and just put them all as a constant size?  You can also try to overlay the tiff image which you did the extraction on.  Note that you may have to rotate/flip your tiff image so it lines up.  There is a function (porespy.tools.align_image_with_openpnm) which can do this for you.

**Reply by nuritg** [reply]
Thanks! It seems to show all pores without scaling.
![image](https://user-images.githubusercontent.com/36815224/183427264-ae5cd19d-9fc4-4003-8239-47c1d7d54de7.png)

 How can I overlay the image?

**Reply by jgostick** [reply]
Just do "file/open/" and navigate to your tiff image.  YOu have to do a few things though...like rotate it before opening, and scale it in the properties pane once it's in paraview.

**Reply by nuritg** [reply]
After opening the image (file- open-AVS UCD Reader), the image doesn't appear. My images are bmp files, is the reader correct? Should I choose something else or open it in a different way?

**Comment by jgostick** [comment]
Try saving it as a TIFF file, using imageio in python, or imagej.  This should open directly.

**Reply by nuritg** [reply]
Thank you! Paraview opened it correctly when using tiff. However, as you said, the image does not align with the network. when should I add that line: (porespy.tools.align_image_with_openpnm)?
![image](https://user-images.githubusercontent.com/36815224/184111584-51c3ba71-65ca-40de-bfff-0c75c9aeeeff.png)

**Reply by jgostick** [reply]
So close! You need to rotate the tiff since paraview and numpy don't agree on axis conventions.  There is a function in porespy.tools ('align_image_with_openpnm') that should do the trick.  You also have to scale the tiff image in paraview.   There are text boxes at the top of the properties panel where you'll but the voxel resolution. I know this needs a video tutorial, but just need to find the time!

**Reply by nuritg** [reply]
It works! However, not all pores appear to be displayed in the pore network.. any idea what can be the reason for that?
Thank you so much for the fast answer each time!
![image](https://user-images.githubusercontent.com/36815224/184538553-93d46341-1423-497b-adad-59da89eb391d.png)

**Reply by jgostick** [reply]
I don't think it's working yet.  The image and the glyphs should line-up much nicer than what is showing.  Maybe a few more flips?  That might also solve the issue of the missing glyphs too.

**Reply by nuritg** [reply]
Is it possible to do several flips using the code itself? if so, how? or where? and do you know how many flips should I make? 
I still didn't find a way to rotate it in Paraview...

**Reply by jgostick** [reply]
I know it's a pain, sorry.  You should do:

1. Extract network from original image using ``snow2``
2. THEN use ``porespy.tools.align_image_to_openpnm`` on the oringal binary image
3. Save the rotated image to a new tiff file using ``imageio.volsave(...)``
4. Save the network using OpenPNM
5. Then open the network and the rotated image in paraview

If this is already what you're doing, then it must mean that the ``align_image_to_openpnm`` function is not doing what it's supposed to. You could try to do the rotations and flips yourself, by trial and error.  You'd use some combination of ``np.flipud, np.fliplr, & np.swapaxes``.

**Reply by nuritg** [reply]
I tried a bit, and it doesn't look good...
but I'll keep trying, and let you know when it works.
